src/Layer/C3DTilesLayer.js

```diff
@@ -3,6 +3,270 @@ import GeometryLayer from 'Layer/GeometryLayer';
+     * @param {string} key - the key of user data to set
+     * @param {object} value - the value of key to set
+     */
+    static setUserData(c3DTileFeatures, key, value) {
+        if (!c3DTileFeatures.length) {
+            return;
+        }
+        // record what layer + tile ids have to be updated
+        const updateStyleInfos = [];
+        c3DTileFeatures.forEach((b) => {
+            const resultFilter = updateStyleInfos.filter(el => el.layer === b.#layer);
+            if (resultFilter.length) {
+                // this layer has already been added
+                const updateStyleInfo = resultFilter[0];
+                if (!updateStyleInfo.tileIds.includes(b._tileId)) {
+                    // tile id of this layer has not been added
+                    updateStyleInfo.tileIds.push(b._tileId);
+                }
+                // already register for an update
+            } else {
+                // register this layer + its tile id
+                updateStyleInfos.push({
+                    layer: b.#layer,
+                    tileIds: [b._tileId],
+                });
+            }
+
+            b.setUserData(key, value, false);
+        });
+
+        // update style
+        updateStyleInfos.forEach((usi) => {
+            usi.layer._styleController.updateStyle(usi.layer, usi.tileIds);
+        });
+    }
+}
+
+class StyleController {
+    constructor() {
+        /** @type {Map<string,THREE.MeshStandardMaterial} */
+        this._fillColorMaterialsBuffer = new Map();
+    }
+
+    /**
+     *
+     * @param {C3DTilesLayer} c3DTilesLayer - C3DTilesLayer to update style
+     * @param {Array<number>|null} [allowTileIdList] - tile ids to allow in updateStyle computation if null all tile id are computed
+     * @returns {boolean} true if style updated false otherwise
+     */
+    updateStyle(c3DTilesLayer, allowTileIdList = null) {
+        if (!c3DTilesLayer._style) {
+            return false;
+        }
+
+        const currentMaterials = [];// list materials used for this update
+        for (const [tileId, c3DTileFeatures] of c3DTilesLayer.tilesC3DTileFeatures) {
+            const tileContent = c3DTilesLayer.object3d.getObjectByProperty('tileId', tileId);
+            if (!tileContent) {
+                throw new Error('no tile content');
+            }
+            tileContent.traverse((child) => {
+                if (child.geometry && child.geometry.attributes._BATCHID) {
+                    // check if tile feature should be updated
+                    if (!(allowTileIdList && !allowTileIdList.includes(tileId))) {
+                        // clear
+                        child.geometry.clearGroups();
+                        child.material = [];
+
+                        // eslint-disable-next-line no-unused-vars
+                        for (const [batchId, c3DTileFeature] of c3DTileFeatures) {
+                            /** @type {THREE.Color} */
```

Comment:
You want to write a function `styleColorToThreeColor` for these lines ?
```
/** @type {THREE.Color} */
let color = null;
if (typeof this._style.fill.color === 'function') {
    color = new THREE.Color(this._style.fill.color(c3DTileFeature));
} else {
    color = new THREE.Color(this._style.fill.color);
}
```
Since this function should be only used in that function ie `updateStyle` it should then be declared in this function scope in order to not pollute C3DTilesLayer API so it will looks like something like this:

```
const styleColorToThreeColor = (feature) => {
 let color = null;
if (typeof this._style.fill.color === 'function') {
    color = new THREE.Color(this._style.fill.color(feature));
} else {
    color = new THREE.Color(this._style.fill.color);
}
return color;
}

...

const color = styleColorToThreeColor(c3DTFeature);
```

Same goes for opacity + material. This is what you mean ?

Reply:
Yes that's what I meant, but I think you can leave the way it is for now. When we will manage more style options I think we should move this style conversion logic into a dedicated `StyleConverter` file but for now it's ok like this